templates/common/_base/files/configure-ovs-network.yaml

```diff
@@ -364,76 +355,57 @@ contents:
       echo "OVS configuration successfully reverted"
     }
 
-    # Reloads NetworkManager if any configuration change was done
-    reload_nm() {
-      if [ $nm_config_changed -eq 0 ]; then
-        # no config was changed, no need to reload
-        return
-      fi
-      nm_config_changed=0
+    # Removes all configuration and reloads NM if necessary
+    rollback_nm() {
+      phys0=$(get_bridge_physical_interface ovs-if-phys0)
+      phys1=$(get_bridge_physical_interface ovs-if-phys1)
       
-      echo "Reloading NetworkManager after configuration changes..."
-
-      # recycle network, so that existing profiles and priorities are re-evaluated
-      nmcli network off
-
-      # wait for no devices to show as connected
-      echo "Waiting for devices to disconnect..."
-      if ! timeout 60 bash -c "while nmcli -g DEVICE,STATE d | grep -v :unmanaged; do sleep 5; done"; then
-        echo "Warning: NetworkManager did not disconnect all devices"
-      fi
+      # Revert changes made by /usr/local/bin/configure-ovs.sh during SDN migration.
+      remove_all_ovn_bridges
       
-      # reload profiles and set networking back on
+      # reload profiles
       nmcli connection reload
-      nmcli network on
-      
-      # restart NetworkManager so that we can wait on `nm-online -s`
-      systemctl restart NetworkManager
 
-      # Wait until all profiles auto-connect
-      echo "Waiting for profiles to activate..."
-      if nm-online -s -t 60; then
-        echo "NetworkManager has activated all suitable profiles after reload"
-      else
-        echo "Warning: NetworkManager has not activated all suitable profiles after reload"
+      # Make sure phys interfaces are up
+      if [ "$phys0" != "" ]; then
+        echo "Waiting for interface $phys0 to activate..."
+        if ! timeout 60 bash -c "while ! nmcli -g DEVICE,STATE c | grep "'"'"$phys0":activated'"'"; do sleep 5; done"; then
+          echo "Warning: $phys0 did not activate"
```

Comment:
In my lab, this fails for the primary interface that was connected to br-ex:
~~~
root@rhel84b ~]# ip r
default via 192.168.130.1 dev br-ex proto dhcp metric 48 
default via 192.168.122.1 dev eth0 proto dhcp metric 101 
default via 192.168.131.1 dev eth6 proto dhcp metric 102 
192.168.122.0/24 dev eth0 proto kernel scope link src 192.168.122.225 metric 101 
192.168.130.0/24 dev br-ex proto kernel scope link src 192.168.130.216 metric 48 
192.168.131.0/24 dev eth6 proto kernel scope link src 192.168.131.206 metric 102 
192.168.140.0/24 dev eth4 proto kernel scope link src 192.168.140.2 metric 104 
[root@rhel84b ~]# nmcli conn
NAME                UUID                                  TYPE           DEVICE 
ovs-if-br-ex        a6cd33f9-58d5-41f7-9852-0c426fcea12a  ovs-interface  br-ex  
Wired connection 1  6a41fbab-57c4-32d4-be67-03cd03d28bbd  ethernet       eth0   
Wired connection 3  2c86a562-6fc4-3fd2-99df-8c524e13765f  ethernet       eth6   
Wired connection 6  152d36ea-290e-3f51-8ec1-7bce2cc4b9a5  ethernet       eth4   
br-ex               38d41220-a29a-48e3-81c5-896c361a9830  ovs-bridge     br-ex  
ovs-if-phys0        97c80a7f-418c-485f-9771-ab16a302c112  ethernet       eth5   
ovs-port-br-ex      1643b0c3-e7df-452d-8194-f15ad1a73191  ovs-port       br-ex  
ovs-port-phys0      53b6a0cf-1f24-48f6-8eaa-d1f14554cae1  ovs-port       eth5   
Wired connection 2  374709a0-9f86-3754-bb69-2773bf8717b9  ethernet       --     
Wired connection 4  00d63637-afd9-3b87-bf90-cae609dee151  ethernet       --     
Wired connection 5  5bc6c5b1-63b6-34e1-839e-e920c7fb8671  ethernet       --     
[root@rhel84b ~]# reboot
Connection to 192.168.140.2 closed by remote host.
Connection to 192.168.140.2 closed.
[root@rhel84 ~]# ssh 192.168.140.2
Activate the web console with: systemctl enable --now cockpit.socket

Register this system with Red Hat Insights: insights-client --register
Create an account or view all your systems at https://red.ht/insights-dashboard
Last login: Wed Apr 27 12:30:48 2022 from 192.168.140.1
[root@rhel84b ~]# ip r
default via 192.168.122.1 dev eth0 proto dhcp metric 100 
default via 192.168.131.1 dev eth6 proto dhcp metric 102 
192.168.122.0/24 dev eth0 proto kernel scope link src 192.168.122.225 metric 100 
192.168.131.0/24 dev eth6 proto kernel scope link src 192.168.131.206 metric 102 
192.168.140.0/24 dev eth4 proto kernel scope link src 192.168.140.2 metric 101 
[root@rhel84b ~]# journalctl --boot | egrep 'while ! nmcli|did not activate'
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: + timeout 60 bash -c 'while ! nmcli -g DEVICE,STATE c | grep "eth5:activated"; do sleep 5; done'
[root@rhel84b ~]# date
Wed Apr 27 12:44:30 EDT 2022
[root@rhel84b ~]# journalctl --boot | egrep 'while ! nmcli|did not activate'
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: + timeout 60 bash -c 'while ! nmcli -g DEVICE,STATE c | grep "eth5:activated"; do sleep 5; done'
Apr 27 12:44:24 rhel84b configure-ovs.sh[1305]: + echo 'Warning: eth5 did not activate'
Apr 27 12:44:24 rhel84b configure-ovs.sh[1305]: Warning: eth5 did not activate
[root@rhel84b ~]# ip r
default via 192.168.122.1 dev eth0 proto dhcp metric 100 
default via 192.168.131.1 dev eth6 proto dhcp metric 102 
192.168.122.0/24 dev eth0 proto kernel scope link src 192.168.122.225 metric 100 
192.168.131.0/24 dev eth6 proto kernel scope link src 192.168.131.206 metric 102 
192.168.140.0/24 dev eth4 proto kernel scope link src 192.168.140.2 metric 101 
[root@rhel84b ~]# ip r
default via 192.168.122.1 dev eth0 proto dhcp metric 100 
default via 192.168.131.1 dev eth6 proto dhcp metric 102 
192.168.122.0/24 dev eth0 proto kernel scope link src 192.168.122.225 metric 100 
192.168.131.0/24 dev eth6 proto kernel scope link src 192.168.131.206 metric 102 
192.168.140.0/24 dev eth4 proto kernel scope link src 192.168.140.2 metric 101 
[root@rhel84b ~]# nmcli conn
NAME                UUID                                  TYPE           DEVICE 
ovs-if-br-ex        2ef3dd64-cdad-465a-ab62-a240b6101e05  ovs-interface  br-ex  
Wired connection 3  2c86a562-6fc4-3fd2-99df-8c524e13765f  ethernet       eth6   
Wired connection 6  152d36ea-290e-3f51-8ec1-7bce2cc4b9a5  ethernet       eth4   
br-ex               c6195ee1-3cb0-4383-bb82-6de56a80a459  ovs-bridge     br-ex  
ovs-if-phys0        4ea6355c-14fb-4927-b522-167a23fc11eb  ethernet       eth0   
ovs-port-br-ex      538b4089-c0c2-4f5c-8c2e-93a1f9e11cf9  ovs-port       br-ex  
ovs-port-phys0      bef9257b-c6d9-44e2-a3b3-300c2e57232e  ovs-port       eth0   
Wired connection 1  6a41fbab-57c4-32d4-be67-03cd03d28bbd  ethernet       --     
Wired connection 2  374709a0-9f86-3754-bb69-2773bf8717b9  ethernet       --     
Wired connection 4  00d63637-afd9-3b87-bf90-cae609dee151  ethernet       --     
[root@rhel84b ~]# ip r
default via 192.168.122.1 dev br-ex proto dhcp metric 48 
default via 192.168.131.1 dev eth6 proto dhcp metric 102 
192.168.122.0/24 dev br-ex proto kernel scope link src 192.168.122.225 metric 48 
192.168.131.0/24 dev eth6 proto kernel scope link src 192.168.131.206 metric 102 
192.168.140.0/24 dev eth4 proto kernel scope link src 192.168.140.2 metric 101 
[root@rhel84b ~]#
~~~

I'm still checking why.

Reply:
The interface initially comes up with ovs-if-phys0:
~~~
[root@rhel84b ~]# journalctl --boot | egrep 'eth5|ovs-if-phys0'
(...)
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: ovs-if-phys0:97c80a7f-418c-485f-9771-ab16a302c112:802-3-ethernet:1651077798:Wed 27 Apr 2022 12\:43\:18 PM EDT:yes:100:no:/org/freedesktop/NetworkManager/Settings/5:yes:eth5:activated:/org/freedesktop/NetworkManager/ActiveConnection/4:ovs-port:/etc/NetworkManager/system-connections/ovs-if-phys0.nmconnection
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: ovs-port-phys0:53b6a0cf-1f24-48f6-8eaa-d1f14554cae1:ovs-port:1651077798:Wed 27 Apr 2022 12\:43\:18 PM EDT:no:0:no:/org/freedesktop/NetworkManager/Settings/3:yes:eth5:activated:/org/freedesktop/NetworkManager/ActiveConnection/6:ovs-bridge:/etc/NetworkManager/system-connections/ovs-port-phys0.nmconnection
(...)
~~~

But then when the connection file is removed, it goes into down and unmanaged state, without a connection. The device never recovers from that:
~~~
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: + rm -f /etc/NetworkManager/system-connections/ovs-if-phys0.nmconnection
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: + echo 'Removed nmconnection file /etc/NetworkManager/system-connections/ovs-if-phys0.nmconnection'
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: Removed nmconnection file /etc/NetworkManager/system-connections/ovs-if-phys0.nmconnection
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:19 'pre-down' [eth5]: new request (0 scripts)
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:19 'pre-down' [eth5]: environment: CONNECTION_DBUS_PATH=/org/freedesktop/NetworkManager/Settings/5
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:19 'pre-down' [eth5]: environment: CONNECTION_FILENAME=/etc/NetworkManager/system-connections/ovs-if-phys0.nmconnection
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:19 'pre-down' [eth5]: environment: CONNECTION_UUID=97c80a7f-418c-485f-9771-ab16a302c112
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:19 'pre-down' [eth5]: environment: CONNECTION_ID=ovs-if-phys0
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:19 'pre-down' [eth5]: environment: DEVICE_IFACE=eth5
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <info>  [1651077804.8246] device (eth5): state change: activated -> deactivating (reason 'removed', sys-iface-state: 'managed')
(...)
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: new request (3 scripts)
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: CONNECTION_DBUS_PATH=/org/freedesktop/NetworkManager/Settings/5
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: CONNECTION_FILENAME=/etc/NetworkManager/system-connections/ovs-if-phys0.nmconnection
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: CONNECTION_UUID=97c80a7f-418c-485f-9771-ab16a302c112
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: CONNECTION_ID=ovs-if-phys0
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: DEVICE_IFACE=eth5
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: DEVICE_IP_IFACE=eth5
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: environment: NM_DISPATCHER_ACTION=down
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: start running ordered scripts...
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5], "/etc/NetworkManager/dispatcher.d/11-dhclient": run script
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.8605] manager: (eth5): assume: cannot generate connection: device has no existing configuration
(...)
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: + echo 'Waiting for interface eth5 to activate...'
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: Waiting for interface eth5 to activate...
Apr 27 12:43:24 rhel84b configure-ovs.sh[1305]: + timeout 60 bash -c 'while ! nmcli -g DEVICE,STATE c | grep "eth5:activated"; do sleep 5; done'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9438] platform: (eth5) signal: link changed: 7: eth5 <UP,LOWER_UP;broadcast,multicast,up,running,lowerup> mtu 1500 arp 1 ethernet? init addrgenmode none addr 52:54:00:5D:C2:86 brd FF:FF:FF:FF:FF:FF driver virtio_net rx:23,2238 tx:27,2800
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9438] device[7c7bb1197c8002cc] (eth5): queued link change for ifindex 7
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9445] device[604f6f7fc4095114] (eth5): add_pending_action (1): 'queued-state-change-disconnected'
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: new request (3 scripts)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9445] device[604f6f7fc4095114] (eth5): queue-state[disconnected, reason:connection-removed, id:474]: queue state change
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: environment: CONNECTION_FILENAME=/etc/NetworkManager/system-connections/ovs-port-phys0.nmconnection
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: environment: CONNECTION_UUID=53b6a0cf-1f24-48f6-8eaa-d1f14554cae1
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: environment: CONNECTION_ID=ovs-port-phys0
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: environment: DEVICE_IFACE=eth5
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: environment: DEVICE_IP_IFACE=eth5
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9445] device[7c7bb1197c8002cc] (eth5): add_pending_action (1): 'queued-state-change-disconnected'
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: environment: PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9445] device[7c7bb1197c8002cc] (eth5): queue-state[disconnected, reason:connection-removed, id:476]: queue state change
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: environment: NM_DISPATCHER_ACTION=down
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: new request (3 scripts)
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: environment: CONNECTION_FILENAME=/etc/NetworkManager/system-connections/ovs-if-phys0.nmconnection
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: environment: CONNECTION_UUID=97c80a7f-418c-485f-9771-ab16a302c112
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: environment: CONNECTION_ID=ovs-if-phys0
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: environment: DEVICE_IFACE=eth5
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: environment: DEVICE_IP_IFACE=eth5
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: environment: PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
Apr 27 12:43:24 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: environment: NM_DISPATCHER_ACTION=down
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9457] device[604f6f7fc4095114] (eth5): queue-state[disconnected, reason:connection-removed, id:474]: change state
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <info>  [1651077804.9457] device (eth5): state change: deactivating -> disconnected (reason 'connection-removed', sys-iface-state: 'managed')
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9457] device[604f6f7fc4095114] (eth5): add_pending_action (2): 'in-state-change'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9457] device[604f6f7fc4095114] (eth5): deactivating device (reason 'connection-removed') [38]
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9457] device[604f6f7fc4095114] (eth5): set metered value 0
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9458] device[604f6f7fc4095114] (eth5): ip4-config: update (commit=1, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9458] device[604f6f7fc4095114] (eth5): ip6-config: update (commit=1, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9458] device[604f6f7fc4095114] (eth5): delete_on_deactivate: schedule cleanup and delete virtual link (id=483)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9461] dispatcher: (29) (eth5) dispatching action 'down'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9463] device[604f6f7fc4095114] (eth5): remove_pending_action (1): 'in-state-change'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9463] device[604f6f7fc4095114] (eth5): remove_pending_action (0): 'queued-state-change-disconnected'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9469] device[7c7bb1197c8002cc] (eth5): queue-state[disconnected, reason:connection-removed, id:476]: change state
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <info>  [1651077804.9469] device (eth5): state change: deactivating -> disconnected (reason 'connection-removed', sys-iface-state: 'managed')
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9469] device[7c7bb1197c8002cc] (eth5): add_pending_action (2): 'in-state-change'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9469] device[7c7bb1197c8002cc] (eth5): will enable userland IPv6LL
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9471] platform-linux: sysctl: setting '/proc/sys/net/ipv6/conf/eth5/disable_ipv6' to '1' (current value is '0')
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9472] platform-linux: sysctl: setting '/proc/sys/net/ipv6/conf/eth5/disable_ipv6' to '0' (current value is '1')
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9473] device[7c7bb1197c8002cc] (eth5): deactivating device (reason 'connection-removed') [38]
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9473] firewalld: [ac7769688c5c4f08,remove*:"eth5"]: firewall zone remove eth5:default (not running, simulate success)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9473] firewalld: [ac7769688c5c4f08,remove*:"eth5"]: complete: drop request simulating success
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9473] platform-linux: sysctl: setting '/proc/sys/net/ipv6/conf/eth5/use_tempaddr' to '0' (current value is identical)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9474] device[7c7bb1197c8002cc] (eth5): ip4-config: update (commit=1, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9474] device[7c7bb1197c8002cc] (eth5): ip4-config: clear IP Config instance (/org/freedesktop/NetworkManager/IP4Config/15)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9475] device[7c7bb1197c8002cc] (eth5): ip6-config: update (commit=1, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9475] device[7c7bb1197c8002cc] (eth5): ip6-config: clear IP Config instance (/org/freedesktop/NetworkManager/IP6Config/15)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9479] device[7c7bb1197c8002cc] (eth5): add_pending_action (3): 'autoactivate'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9479] dispatcher: (31) (eth5) dispatching action 'down'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9482] device[7c7bb1197c8002cc] (eth5): remove_pending_action (2): 'in-state-change'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9482] device[7c7bb1197c8002cc] (eth5): remove_pending_action (1): 'queued-state-change-disconnected'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9488] device[604f6f7fc4095114] (eth5): delete_on_deactivate: cleanup and delete virtual link (id=483)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9488] device[604f6f7fc4095114] (eth5): unrealize (ifindex 0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9489] device[604f6f7fc4095114] (eth5): unmanaged: flags set to [platform-init,!sleeping,!by-type,!user-explicit,!user-settings=0x10/0x79/unmanaged/unrealized], set-unmanaged [platform-init=0x10])
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9490] device[604f6f7fc4095114] (eth5): unmanaged: flags set to [platform-init,!sleeping,!user-settings=0x10/0x51/unmanaged/unrealized], forget [parent,by-type,user-explicit,user-udev,external-down,is-slave=0xe2c])
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <info>  [1651077804.9490] device (eth5): state change: disconnected -> unmanaged (reason 'user-requested', sys-iface-state: 'managed')
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9490] device[604f6f7fc4095114] (eth5): add_pending_action (1): 'in-state-change'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9490] device[604f6f7fc4095114] (eth5): taking down device 0
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9491] device[604f6f7fc4095114] (eth5): remove_pending_action (0): 'in-state-change'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9491] device[604f6f7fc4095114] (eth5): add_pending_action (1): 'autoactivate'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9495] manager: (eth5): assume: cannot generate connection: device has no existing configuration
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9495] device[7c7bb1197c8002cc] (eth5): remove_pending_action (0): 'autoactivate'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9497] manager: (eth5): removing device (managed 0, wol 0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9497] device[604f6f7fc4095114] (eth5): ip4-config: update (commit=0, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9497] device[604f6f7fc4095114] (eth5): ip6-config: update (commit=0, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9497] device[604f6f7fc4095114] (eth5): remove_pending_action (0): 'autoactivate'
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9498] device[604f6f7fc4095114] (eth5): disposing
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9498] device[604f6f7fc4095114] (eth5): ip4-config: update (commit=1, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9498] device[604f6f7fc4095114] (eth5): ip6-config: update (commit=1, new-config=d2791a87cbe20953, route-table-sync-mode=0)
Apr 27 12:43:24 rhel84b NetworkManager[1017]: <debug> [1651077804.9498] device[604f6f7fc4095114] (eth5): finalize(): NMDeviceOvsPort
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:20 'down' [eth5]: completed (3 scripts)
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:22 'pre-up' [eth5]: start running ordered scripts...
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:22 'pre-up' [eth5], "/etc/NetworkManager/dispatcher.d/pre-up.d/10-ofport-request.sh": run script
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:22 'pre-up' [eth5], "/etc/NetworkManager/dispatcher.d/pre-up.d/10-ofport-request.sh": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:22 'pre-up' [eth5]: completed (1 scripts)
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: start running ordered scripts...
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5], "/etc/NetworkManager/dispatcher.d/11-dhclient": run script
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5], "/etc/NetworkManager/dispatcher.d/11-dhclient": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": run script
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": run script
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:29 'down' [eth5]: completed (3 scripts)
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: start running ordered scripts...
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5], "/etc/NetworkManager/dispatcher.d/11-dhclient": run script
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5], "/etc/NetworkManager/dispatcher.d/11-dhclient": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": run script
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": run script
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": complete
Apr 27 12:43:25 rhel84b nm-dispatcher[1042]: req:31 'down' [eth5]: completed (3 scripts)
Apr 27 12:44:24 rhel84b configure-ovs.sh[1305]: + echo 'Warning: eth5 did not activate'
Apr 27 12:44:24 rhel84b configure-ovs.sh[1305]: Warning: eth5 did not activate
~~~
~~~

Reply:
[journal.txt](https://github.com/openshift/machine-config-operator/files/8574879/journal.txt)

Reply:
The sequence for me to trigger this is:
~~~
rm -f /run/configure-ovs-boot-done
rm -f /etc/ovnk/iface_default_hint
rm -f /var/lib/ovnk/iface_default_hint
rm -f /etc/ovnk/ -R
rm -f /etc/NetworkManager/system-connections/*
eth4=$(nmcli -f uuid,device conn | awk '/eth4/ {print $1}')
nmcli conn mod $eth4 ipv4.method manual ipv4.address 192.168.140.2/24
nmcli conn up $eth4
systemctl restart NetworkManager


systemctl restart ovs-configure
~~~

After this, everything looks great and set up correctly, so then I reboot:
~~~
reboot
~~~

And that triggers the issue.

Reply:
What solves this for me is:
~~~
rollback_nm() {
  phys0=$(get_bridge_physical_interface ovs-if-phys0)
  phys1=$(get_bridge_physical_interface ovs-if-phys1)

  # Revert changes made by /usr/local/bin/configure-ovs.sh during SDN migration.
  remove_all_ovn_bridges

  # reload profiles
  nmcli connection reload

  # Make sure phys interfaces are up
  if [ "$phys0" != "" ]; then
    local state=$(nmcli -f DEVICE,STATE,CONNECTION dev | awk '$1=="'${phys0}'" {print $2}')
    local connection=$(nmcli -f DEVICE,STATE,CONNECTION dev | awk '$1=="'${phys0}'" {print $3}')
    if [ "${state}" == "disconnected" ] && [ "${connection}" == "--" ]; then
        nmcli dev connect "${phys0}"
    fi
    echo "Waiting for interface $phys0 to activate..."
    if ! timeout 60 bash -c "while ! nmcli -g DEVICE,STATE c | grep "'"'"$phys0":activated'"'"; do sleep 5; done"; then
      echo "Warning: $phys0 did not activate"
    fi
  fi
  if [ "$phys1" != "" ]; then
    local state=$(nmcli -f DEVICE,STATE,CONNECTION dev | awk '$1=="'${phys1}'" {print $2}')
    local connection=$(nmcli -f DEVICE,STATE,CONNECTION dev | awk '$1=="'${phys1}'" {print $3}')
    if [ "${state}" == "disconnected" ] && [ "${connection}" == "--" ]; then
        nmcli dev connect "${phys1}"
    fi
    echo "Waiting for interface $phys1 to activate..."
    if ! timeout 60 bash -c "while ! nmcli -g DEVICE,STATE c | grep "'"'"$phys1":activated'"'"; do sleep 5; done"; then
      echo "Warning: $phys1 did not activate"
    fi
  fi
}
~~~

Reply:
With the caveat that now, I have a new "eth0" connection:
~~~
[root@rhel84b ~]# nmcli conn
NAME                UUID                                  TYPE           DEVICE 
ovs-if-br-ex        36fc031f-e988-4b31-ad19-4927e922e216  ovs-interface  br-ex  
Wired connection 2  9bb815cb-ea3e-3981-877d-61a7acbd80a1  ethernet       eth5   
Wired connection 3  2c86a562-6fc4-3fd2-99df-8c524e13765f  ethernet       eth6   
Wired connection 6  152d36ea-290e-3f51-8ec1-7bce2cc4b9a5  ethernet       eth4   
br-ex               e771380b-c4ca-4313-9205-76bc8039c54c  ovs-bridge     br-ex  
ovs-if-phys0        8f863d17-d502-46ed-9148-fc352f9e210a  ethernet       eth0   
ovs-port-br-ex      34396932-07c5-4fbe-9cc2-f8110fa71725  ovs-port       br-ex  
ovs-port-phys0      a1d07265-c091-4682-ba7b-d9a4b36c8e6b  ovs-port       eth0   
eth0                23db7bcf-7df7-4289-9bd0-ddf8186f16b0  ethernet       --     
Wired connection 1  2445efcf-49eb-3044-ac2b-314693b58d32  ethernet       --     
Wired connection 4  00d63637-afd9-3b87-bf90-cae609dee151  ethernet       --     
[root@rhel84b ~]# journalctl --boot | egrep 'local state|local connection' -C5
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:21 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": run script
Apr 27 13:21:52 rhel84b configure-ovs.sh[1319]: ++ nmcli -f DEVICE,STATE,CONNECTION dev
Apr 27 13:21:52 rhel84b configure-ovs.sh[1319]: ++ awk '$1=="eth0" {print $2}'
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:21 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": complete
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:21 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": run script
Apr 27 13:21:52 rhel84b configure-ovs.sh[1319]: + local state=disconnected
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:21 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": complete
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:21 'down' [br-ex]: completed (3 scripts)
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:22 'pre-up' [eth0]: start running ordered scripts...
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:22 'pre-up' [eth0], "/etc/NetworkManager/dispatcher.d/pre-up.d/10-ofport-request.sh": run script
Apr 27 13:21:52 rhel84b NetworkManager[1029]: <debug> [1651080112.5675] dispatcher: (21) /etc/NetworkManager/dispatcher.d/11-dhclient succeeded
--
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:23 'pre-up' [br-ex]: completed (1 scripts)
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:29 'down' [br-ex]: start running ordered scripts...
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:29 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/11-dhclient": run script
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:29 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/11-dhclient": complete
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:29 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": run script
Apr 27 13:21:52 rhel84b configure-ovs.sh[1319]: + local connection=--
Apr 27 13:21:52 rhel84b configure-ovs.sh[1319]: + '[' disconnected == disconnected ']'
Apr 27 13:21:52 rhel84b configure-ovs.sh[1319]: + '[' -- == -- ']'
Apr 27 13:21:52 rhel84b configure-ovs.sh[1319]: + nmcli dev connect eth0
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:29 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/20-chrony-dhcp": complete
Apr 27 13:21:52 rhel84b nm-dispatcher[1054]: req:29 'down' [br-ex], "/etc/NetworkManager/dispatcher.d/20-chrony-onoffline": run script
[root@rhel84b ~]# 
~~~

Reply:
And the commands to check the state (instead of the list with fields) would better be something like:
~~~
[root@rhel84b ~]# nmcli --get-values GENERAL.STATE dev show eth0
30 (disconnected)
[root@rhel84b ~]# nmcli --get-values GENERAL.CONNECTION dev show eth0

[root@rhel84b ~]# 
~~~

Reply:
Perhaps that's because they are auto-generated profiles that live in `/run`. Once `ovs-if-phys0` profile exists, they are no longer generated. And when `ovs-if-phys0` is removed, there's no actual profile to activate it with.

Once there, we could for example do `nmcli device connect eth5`. But is not that good because
* NM auto-generates a profile for eth5, but this time in `/etc` instead of `/run`. That has the effect of persisting something that should be ephemeral. I don't know if this is a bug or not in NM. WE can probably workaround it.
* We cannot do that for virtual devices like bonds because when disconnected NM does not know about them. But we can probably assume these would have an explicit profile to be connected with.

I will see if I can work something out.

Reply:
Pushed some workaround to this problem